processing/src/main/java/org/apache/druid/math/expr/Function.java

```diff
@@ -3742,23 +3746,95 @@ public ExpressionType getOutputType(Expr.InputBindingInspector inspector, List<E
     @Override
     Expr getScalarArgument(List<Expr> args)
     {
-      return args.get(0);
+      return args.get(SCALAR_ARG);
     }
 
     @Override
     Expr getArrayArgument(List<Expr> args)
     {
-      return args.get(1);
+      return args.get(ARRAY_ARG);
     }
 
     @Override
-    ExprEval doApply(ExprEval arrayExpr, ExprEval scalarExpr)
+    ExprEval doApply(ExprEval arrayEval, ExprEval scalarEval)
     {
-      final Object[] array = arrayExpr.castTo(scalarExpr.asArrayType()).asArray();
+      final Object[] array = arrayEval.asArray();
       if (array == null) {
         return ExprEval.ofLong(null);
       }
-      return ExprEval.ofLongBoolean(Arrays.asList(array).contains(scalarExpr.value()));
+
+      if (scalarEval.value() == null) {
+        return Arrays.asList(array).contains(null) ? ExprEval.ofLongBoolean(true) : ExprEval.ofLong(null);
```

Comment:
I wonder if the `array` contains `null` - then doesn't all places when `false` would have been returned should return `null`? for example:

```
c IN (2,null)
c = 2 OR c = null
c = 2 OR null
```
this last rewrite essentially mean that `false` is no more
....other way around is to think about `=` as `IS NOT DISTINCT FROM` ; however in that case the return value will never be `null` as `<=>` is a 2-valued

this is not a blocking comment; I was just wondering...

Reply:
The idea is that if the array contains `null` then the comparison acts like `IS NOT DISTINCT FROM` (always returns `true` or `false`), whereas if the array does not contain `null`, then the comparison acts like `=` (returns `null` if the lhs is `null`). It's the same way the native `in` and `inType` filters behave, so this is designed to align with those.

A future patch would convert from SQL `IN` to this `SCALAR_IN_ARRAY` function. We'll need to make sure to handle `NULL` appropriately in that patch-- it would not be ok to rewrite `c IN (2, NULL)` to `SCALAR_IN_ARRAY(c, ARRAY[2, NULL])`.